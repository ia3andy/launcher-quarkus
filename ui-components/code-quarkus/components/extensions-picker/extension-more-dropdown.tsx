import React, { useEffect, useState } from 'react';
import { useAnalytics } from '@quarkusio/code-quarkus.core.analytics';
import { CopyToClipboard } from '@quarkusio/code-quarkus.core.components';
import { ExtensionEntry } from './extensions-picker';
import { FaAngleDown, FaAngleUp, FaMap } from 'react-icons/fa';
import DropdownItem from 'react-bootstrap/DropdownItem';
import { Button, Dropdown } from 'react-bootstrap';

interface ExtensionMoreDropdownProps extends ExtensionEntry {
  active: boolean;
}

export function ExtensionMoreDropdown(props: ExtensionMoreDropdownProps) {
  const [ isOpen , setIsOpen ] = useState(false);
  const analytics = useAnalytics();
  const gav = `${props.id}:${props.version}`;
  const gavArray = gav.split(':');
  const bomArray = props.bom && props.bom.split(':');
  const addMvnExt = `./mvnw quarkus:add-extension -Dextensions="${props.id}"`;
  const addCliExt = `quarkus ext add ${props.id}`;
  const addGradleExt = `./gradlew addExtension --extensions="${props.id}"`;
  const pomDepXml = `        <dependency>\n            <groupId>${gavArray[0]}</groupId>\n            <artifactId>${gavArray[1]}</artifactId>\n        </dependency>`;
  const pomBomXml = bomArray && `        <dependency>\n            <groupId>${bomArray[0]}</groupId>\n            <artifactId>${bomArray[1]}</artifactId>\n            <version>${bomArray[2]}</version>\n            <type>pom</type>\n            <scope>import</scope>\n        </dependency>`;

  const openGuide = () => {
    analytics.event('Extension', 'Click "Open Extension Guide" link', props.id);
  };

  useEffect(() => {
    if(!props.active && isOpen) {
      setIsOpen(false);
    }
  }, [ props.active, isOpen, setIsOpen ]);

  return (
    <Dropdown onClick={(e) => e.stopPropagation()} onToggle={setIsOpen} show={isOpen}>
      <Dropdown.Toggle className="extension-more-button">
        {isOpen ? <FaAngleUp /> : <FaAngleDown />}
      </Dropdown.Toggle>
      {props.active && (
        <Dropdown.Menu
          align="right"
          onClick={(e) => e.stopPropagation()}
        >
          <DropdownItem key="cli" as={Button}>
            <CopyToClipboard id="more-copy-cli"
              event={[ 'Extension', 'Copy the Quarkus CLI add extension command', props.id ]}
              content={addCliExt}
              tooltipPlacement="left" zIndex={201}
            >Add it to your project with the Quarkus CLI</CopyToClipboard>
          </DropdownItem>
          <DropdownItem key="maven" as={Button}>
            <CopyToClipboard id="more-copy-maven"
              event={[ 'Extension', 'Copy the Maven add extension command', props.id ]}
              content={addMvnExt}
              tooltipPlacement="left" zIndex={201}
            >Add it to your project with Maven</CopyToClipboard>
          </DropdownItem>
          <DropdownItem key="gradle" as={Button}>
            <CopyToClipboard id="more-copy-gradle"
              event={[ 'Extension', 'Copy the Gradle add extension command', props.id ]}
              content={addGradleExt} tooltipPlacement="left" zIndex={201}
            >Add it to your project with Gradle</CopyToClipboard>
          </DropdownItem>
          <DropdownItem key="xml" as={Button}>
            <CopyToClipboard id="more-copy-pom" event={[ 'Extension', 'Copy the Maven pom.xml dependency snippet', props.id ]}
              content={pomDepXml}
              tooltipPlacement="left" zIndex={201}
            >Get the Maven pom.xml dependency snippet</CopyToClipboard>
          </DropdownItem>
          {pomBomXml && (
            <DropdownItem key="xml" as={Button}>
              <CopyToClipboard id="more-copy-pom" event={[ 'Extension', 'Copy the Maven pom.xml bom snippet', props.id ]}
                content={pomBomXml}
                tooltipPlacement="left" zIndex={201}
              >Get the Maven pom.xml bom snippet</CopyToClipboard>
            </DropdownItem>
          )}          
          <DropdownItem key="id" as={Button}>
            <CopyToClipboard id="more-copy-gav" event={[ 'Extension', 'Copy the extension GAV', props.id ]} content={gav}
              tooltipPlacement="left"
              zIndex={201}>Get the groupId:artifactId:version</CopyToClipboard>
          </DropdownItem>
          {props.guide && (<DropdownItem key="guide" href={props.guide} target="_blank" onClick={openGuide}>
            <FaMap/> See the extension guide
          </DropdownItem>)}
        </Dropdown.Menu>)
      }
    </Dropdown>

  );
}
