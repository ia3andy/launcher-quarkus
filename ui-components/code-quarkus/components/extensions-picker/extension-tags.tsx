import React from 'react';

export function ExtensionTags(props: { status?: string }) {
  if (!props.status) {
    return <React.Fragment/>;
  }

  switch (props.status) {
  case 'preview':
    return (<span
      className="extension-tag preview"
      title="This is work in progress. API or configuration properties might change as the extension matures. Give us your feedback :)"
    >PREVIEW</span>);
  case 'experimental':
    return (<span
      className="extension-tag experimental"
      title="Early feedback is requested to mature the idea. There is no guarantee of stability nor long term presence in the platform until the solution matures."
    >EXPERIMENTAL</span>);
  case 'deprecated':
    return (<span
      title="This extension has been deprecated. It is likely to be replaced or removed in a future version of Quarkus."
      className="extension-tag deprecated"
    >DEPRECATED</span>);
  case 'code':
    return (<span
      title="This extension provides starter code (may not be available in all languages)."
      className="extension-tag code"
    >CODE</span>);
  default:
    return (<span
      className="extension-tag custom"
    >{props.status.toUpperCase()}</span>);
  }
}