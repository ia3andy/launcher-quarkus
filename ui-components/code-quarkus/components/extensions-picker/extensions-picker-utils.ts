import { ExtensionEntry } from './extensions-picker';

export const filterFunction = (filter: string) => (d: ExtensionEntry) => {
  const filterLowerCase = filter.trim().toLowerCase();
  if (!filterLowerCase) {
    return true;
  }
  const shortName = d.shortName ? d.shortName.toLowerCase() : '';
  if (filterLowerCase === shortName) {
    return true;
  }
  return d.name.toLowerCase().includes(filterLowerCase)
    || d.keywords.filter(l => l.startsWith(filterLowerCase)).length > 0
    || (d.category && d.category.toLowerCase().startsWith(filterLowerCase))
    || d.tags.filter(l => l.startsWith(filterLowerCase)).length > 0
    || shortName.startsWith(filterLowerCase);
};

export const sortFunction = (filter: string) => (a: ExtensionEntry, b: ExtensionEntry) => {
  const filterLowerCase = filter.trim().toLowerCase();
  if (!filterLowerCase) {
    return a.order > b.order ? 1 : -1;
  }
  const startWithAShortName = !!a.shortName && a.shortName.toLowerCase().startsWith(filterLowerCase);
  const startWithBShortName = !!b.shortName && b.shortName.toLowerCase().startsWith(filterLowerCase);
  if (startWithAShortName !== startWithBShortName) {
    return startWithAShortName ? -1 : 1;
  }
  const startWithOneOfATags = a.tags.filter(l => l.startsWith(filterLowerCase)).length > 0;
  const startWithOneOfBTags = b.tags.filter(l => l.startsWith(filterLowerCase)).length > 0;
  if (startWithOneOfATags !== startWithOneOfBTags) {
    return startWithOneOfATags ? -1 : 1;
  }
  const startWithOneOfAKeywords = a.keywords.filter(l => l.startsWith(filterLowerCase)).length > 0;
  const startWithOneOfBKeywords = b.keywords.filter(l => l.startsWith(filterLowerCase)).length > 0;
  if (startWithOneOfAKeywords !== startWithOneOfBKeywords) {
    return startWithOneOfAKeywords ? -1 : 1;
  }
  if (a.name.toLowerCase().startsWith(filterLowerCase) !== b.name.toLowerCase().startsWith(filterLowerCase)) {
    return a.name.toLowerCase().startsWith(filterLowerCase) ? -1 : 1;
  }
  return a.order > b.order ? 1 : -1;
};

export const removeDuplicateIds = (filter: string, entries: ExtensionEntry[]): ExtensionEntry[] => {
  if (!filter) {
    return entries;
  }
  const ids = new Set();
  return entries
    .filter(e => {
      if (ids.has(e.id)) {
        return false;
      }
      ids.add(e.id);
      return true;
    });
};

export const processEntries = (filter: string, entries: ExtensionEntry[]): ExtensionEntry[] => {
  return removeDuplicateIds(filter, entries.filter(filterFunction(filter)).sort(sortFunction(filter)));
};
