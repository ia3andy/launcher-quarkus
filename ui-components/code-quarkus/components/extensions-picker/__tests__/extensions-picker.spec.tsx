import { cleanup, fireEvent, render } from '@testing-library/react';
import React from 'react';
import { ExtensionEntry, ExtensionsPicker } from '../extensions-picker';
import { filterFunction, sortFunction } from '../extensions-picker-utils';
import { act } from 'react-dom/test-utils';

afterEach(() => {
  cleanup();
});

const entries: ExtensionEntry[] = [
  {
    'id': 'io.quarkus:quarkus-arc',
    'name': 'ArC',
    'version': 'test-version',
    'tags': [],
    'keywords': [
      'arc',
      'cdi',
      'dependency-injection',
      'di',
      'label'
    ],
    'default': false,
    'description': 'Build time CDI dependency injection',
    'shortName': 'CDI',
    'category': 'Core',
    'order': 0,
  },
  {
    'id': 'io.quarkus:quarkus-camel-netty4-http',
    'version': 'test-version',
    'name': 'Camel Netty4 test HTTP',
    'tags': [ 'preview' ],
    'default': false,
    'keywords': [
      'camel-netty4-http',
      'camel'
    ],
    'description': 'Camel support for Netty',
    'category': 'Integration',
    'order': 2,
  },
  {
    'id': 'some-id',
    'version': 'test-version',
    'name': 'A CDI in name test',
    'tags': [ 'experimental' ],
    'default': false,
    'keywords': [
      'lambda',
      'amazon-lambda',
      'aws-lambda',
      'amazon',
      'aws',
      'label'
    ],
    'shortName': 'a shortname',
    'description': 'Some description',
    'category': 'Cloud',
    'order': 1,
  },
];

describe('<ExtensionsPicker />', () => {

  it('renders the ExtensionsPicker correctly', () => {
    const setFilter = jest.fn();
    const comp = render(<ExtensionsPicker placeholder="" entries={entries} value={{ extensions: [] }} onChange={() => { }} buildTool="MAVEN" filter="" setFilter={setFilter}  />);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('show results for valid search', async () => {
    const setFilter = jest.fn();
    const comp = render(<ExtensionsPicker placeholder="" entries={entries} value={{ extensions: [] }} onChange={() => { }} buildTool="MAVEN" filter="" setFilter={setFilter} />);

    const searchField = comp.getByLabelText('Search extensions');
    fireEvent.change(searchField, { target: { value: 'CDI' } });
    expect(setFilter).toBeCalledTimes(1);
    expect(setFilter).lastCalledWith('CDI');
    act(() => {
      comp.rerender(<ExtensionsPicker placeholder="" entries={entries} value={{ extensions: [] }} onChange={() => { }} buildTool="MAVEN" filter="CDI" setFilter={setFilter} />);
    });
    const result = await comp.findAllByText(entries[0].description!);
    expect((result as HTMLElement[]).length).toBe(1);
    expect(comp.asFragment()).toMatchSnapshot();
  });

  it('select values and save', async () => {
    const setFilter = jest.fn();
    const handleChange = jest.fn();
    const comp = render(<ExtensionsPicker placeholder="" entries={entries} value={{ extensions: [] }} onChange={handleChange} buildTool="MAVEN" filter="" setFilter={setFilter} />);

    const searchField = comp.getByLabelText('Search extensions');
    fireEvent.change(searchField, { target: { value: 'netty' } });
    expect(setFilter).toBeCalledTimes(1);
    expect(setFilter).lastCalledWith('netty');
    act(() => {
      comp.rerender(<ExtensionsPicker placeholder="" entries={entries} value={{ extensions: [] }} onChange={() => { }} buildTool="MAVEN" filter="netty" setFilter={setFilter} />);
    });
    const item = await comp.findAllByText(entries[1].description!);
    fireEvent.click(item[0]);
    expect(comp.asFragment()).toMatchSnapshot();
  });
});

describe('filterFunction', () => {
  it('when not using a filter, all extensions should be returned', () => {
    expect(entries.filter(filterFunction('')))
      .toEqual(entries);
  });

  it('when using exact shortname, the result should include the other corresponding extensions too', () => {
    expect(entries.filter(filterFunction('cdi')))
      .toEqual([ entries[0], entries[2] ]);
  });

  it('when using start of shortname, the result should include the other corresponding extensions too', () => {
    expect(entries.filter(filterFunction('cd')))
      .toEqual([ entries[0], entries[2] ]);
  });

  it('when using part of name, it should return the corresponding extensions', () => {
    expect(entries.filter(filterFunction('test')))
      .toEqual([ entries[1], entries[2] ]);
  });

  it('when using label, it should return the corresponding extensions', () => {
    expect(entries.filter(filterFunction('label')))
      .toEqual([ entries[0], entries[2] ]);
  });

  it('when using part of label, it should return the corresponding extensions', () => {
    expect(entries.filter(filterFunction('labe')))
      .toEqual([ entries[0], entries[2] ]);
  });

  it('when using part of category (not start), it should not return it', () => {
    expect(entries.filter(filterFunction('oud')))
      .toEqual([]);
  });

  it('when using start of category, it should return it', () => {
    expect(entries.filter(filterFunction('clou')))
      .toEqual([ entries[2] ]);
  });

  it('when using start of tag, it should return it', () => {
    expect(entries.filter(filterFunction('prev')))
      .toEqual([ entries[1] ]);
  });
});

describe('sortFunction', () => {
  it('when using not filter, it should use the order field', () => {
    const sorted = entries.slice(0).sort(sortFunction(''));
    expect(sorted).toEqual([ entries[0], entries[2], entries[1] ]);
  });

  it('when using start of shortname of an extension, it should be first', () => {
    expect(sortFunction('cdi')(entries[0], entries[2]))
      .toEqual(-1);
    expect(sortFunction('cdi')(entries[2], entries[0]))
      .toEqual(1);
  });

  it('when using one of the label of an extension, it should be first', () => {
    expect(sortFunction('amazon')(entries[2], entries[0]))
      .toEqual(-1);
    expect(sortFunction('amazon')(entries[0], entries[2]))
      .toEqual(1);
  });

  it('when using one part of the label of an extension, it should be first', () => {
    expect(sortFunction('amazo')(entries[2], entries[0]))
      .toEqual(-1);
    expect(sortFunction('amazo')(entries[0], entries[2]))
      .toEqual(1);
  });

  it('when using start of the name of an extension, it should be first', () => {
    expect(sortFunction('a')(entries[0], entries[1]))
      .toEqual(-1);
    expect(sortFunction('a')(entries[1], entries[0]))
      .toEqual(1);
  });

  it('when using start of the tags of an extension, it should be first', () => {
    expect(sortFunction('prev')(entries[1], entries[0]))
      .toEqual(-1);
    expect(sortFunction('prev')(entries[0], entries[1]))
      .toEqual(1);
  });

  it('when no match, it should compare the order', () => {
    const sorted = entries.slice(0).sort(sortFunction('anomatch'));
    expect(sorted).toEqual([ entries[0], entries[2], entries[1] ]);
  });

});