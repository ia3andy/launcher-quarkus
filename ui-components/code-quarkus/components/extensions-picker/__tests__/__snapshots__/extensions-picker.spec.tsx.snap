// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExtensionsPicker /> renders the ExtensionsPicker correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="Extensions picker"
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="search-bar responsive-container"
      >
        <div
          class="form-group"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <input
            aria-label="Search extensions"
            autocomplete="off"
            class="extensions-search-input form-control"
            id="extensions-search-input"
            placeholder=""
            type="search"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="main-container responsive-container"
    >
      <div
        class="extension-list-wrapper"
      >
        <div
          class="extension-category"
        >
          Core
        </div>
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch io.quarkus:quarkus-arc extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="ArC (test-version)"
            >
              ArC
            </span>
          </div>
          <div
            class="extension-description"
            title="Build time CDI dependency injection"
          >
            Build time CDI dependency injection
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="extension-category"
        >
          Cloud
        </div>
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch some-id extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="A CDI in name test (test-version)"
            >
              A CDI in name test
            </span>
            <span
              class="extension-tag experimental"
              title="Early feedback is requested to mature the idea. There is no guarantee of stability nor long term presence in the platform until the solution matures."
            >
              EXPERIMENTAL
            </span>
          </div>
          <div
            class="extension-description"
            title="Some description"
          >
            Some description
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="extension-category"
        >
          Integration
        </div>
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch io.quarkus:quarkus-camel-netty4-http extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="Camel Netty4 test HTTP (test-version)"
            >
              Camel Netty4 test HTTP
            </span>
            <span
              class="extension-tag preview"
              title="This is work in progress. API or configuration properties might change as the extension matures. Give us your feedback :)"
            >
              PREVIEW
            </span>
          </div>
          <div
            class="extension-description"
            title="Camel support for Netty"
          >
            Camel support for Netty
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ExtensionsPicker /> select values and save 1`] = `
<DocumentFragment>
  <div
    aria-label="Extensions picker"
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="search-bar responsive-container"
      >
        <div
          class="form-group"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <input
            aria-label="Search extensions"
            autocomplete="off"
            class="extensions-search-input form-control"
            id="extensions-search-input"
            placeholder=""
            type="search"
            value="netty"
          />
        </div>
      </div>
    </div>
    <div
      class="main-container responsive-container"
    >
      <div
        class="extension-search-clear"
      >
        Search results (
        <a
          class="btn btn-primary"
        >
          Clear search
        </a>
        )
      </div>
      <div
        class="extension-list-wrapper"
      >
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch io.quarkus:quarkus-camel-netty4-http extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="Camel Netty4 test HTTP (test-version)"
            >
              Camel Netty4 test HTTP
            </span>
            <span
              class="extension-tag preview"
              title="This is work in progress. API or configuration properties might change as the extension matures. Give us your feedback :)"
            >
              PREVIEW
            </span>
          </div>
          <div
            class="extension-description"
            title="Camel support for Netty"
          >
            Camel support for Netty
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<ExtensionsPicker /> show results for valid search 1`] = `
<DocumentFragment>
  <div
    aria-label="Extensions picker"
    class="extensions-picker"
  >
    <div
      class="control-container"
    >
      <div
        class="search-bar responsive-container"
      >
        <div
          class="form-group"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <input
            aria-label="Search extensions"
            autocomplete="off"
            class="extensions-search-input form-control"
            id="extensions-search-input"
            placeholder=""
            type="search"
            value="CDI"
          />
        </div>
      </div>
    </div>
    <div
      class="main-container responsive-container"
    >
      <div
        class="extension-search-clear"
      >
        Search results (
        <a
          class="btn btn-primary"
        >
          Clear search
        </a>
        )
      </div>
      <div
        class="extension-list-wrapper"
      >
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch io.quarkus:quarkus-arc extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="ArC (test-version)"
            >
              ArC
            </span>
          </div>
          <div
            class="extension-description"
            title="Build time CDI dependency injection"
          >
            Build time CDI dependency injection
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="extension-row"
        >
          <div
            aria-label="Switch some-id extension"
            class="extension-selector"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
              />
            </svg>
          </div>
          <div
            class="extension-summary"
          >
            <span
              class="extension-name"
              title="A CDI in name test (test-version)"
            >
              A CDI in name test
            </span>
            <span
              class="extension-tag experimental"
              title="Early feedback is requested to mature the idea. There is no guarantee of stability nor long term presence in the platform until the solution matures."
            >
              EXPERIMENTAL
            </span>
          </div>
          <div
            class="extension-description"
            title="Some description"
          >
            Some description
          </div>
          <div
            class="extension-more"
          >
            <div
              class="dropdown"
            >
              <button
                aria-expanded="false"
                aria-haspopup="true"
                class="extension-more-button dropdown-toggle btn btn-primary"
                type="button"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 320 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
